### PR TITLE
plots: add plot markers to DVC files

### PR DESCRIPTION
**Pull request by efiop**
* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here. If the CLI API is changed, I have updated [tab completion scripts](https://github.com/iterative/dvc/tree/master/scripts/completion).

https://github.com/iterative/dvc.org/issues/1255

* [x] ❌ I will check DeepSource, CodeClimate, and other sanity checks below. (We consider them recommendatory and don't expect everything to be addressed. Please fix things that actually improve code or fix bugs.)

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

File: dvc/repo/plots/show.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        for rev in repo.brancher(revs=revs):
+            plots.update(_collect_plots(repo))
+
+    if not plots:
+        datafile, plot = _show(repo, datafile=None, revs=revs, **kwargs)
+        return {datafile: plot}
+
+    ret = {}
+    for plot in plots:
+        ret[plot] = _show(repo, datafile=plot, revs=revs, **kwargs)[1]
```
Comment (efiop) [inline dvc/repo/plots/show.py:203]: need to catch errors and plot whatever we can.

File: dvc/stage/params.py [hunk 2]
```diff
@@ -11,12 +11,16 @@ class StageParams:
     PARAM_META = "meta"
     PARAM_ALWAYS_CHANGED = "always_changed"
     PARAM_PARAMS = "params"
+    PARAM_METRICS = "metrics"
+    PARAM_PLOTS = "plots"
 
 
 class OutputParams(Enum):
```
Comment (skshetry) [inline dvc/stage/params.py:18]: Still needed?

File: dvc/serialize.py [hunk 3]
```diff
@@ -92,7 +91,20 @@ def to_pipeline_file(stage: "PipelineStage"):
...
+            ),
+        ),
+        (
+            stage.PARAM_METRICS,
+            _get_outs([out for out in stage.outs if out.metric]),
+        ),
+        (
+            stage.PARAM_PLOTS,
+            _get_outs([out for out in stage.outs if out.plot]),
+        ),
```
Comment (skshetry) [inline dvc/serialize.py:107]: Let's do this in a single loop in `_get_outs()`. What do you say?

File: dvc/stage/params.py (hunk 2 above)
Comment (efiop) [reply dvc/stage/params.py:18]: OutputParams? Not really, I just didn't touch them in this PR for now.

File: dvc/command/run.py [hunk 4]
```diff
@@ -140,6 +144,20 @@ def add_parser(subparsers, parent_parser):
...
+        action="append",
+        default=[],
+        help="Declare output plot file.",
+        metavar="<path>",
+    )
+    run_parser.add_argument(
+        "--plots-no-cache",
+        action="append",
+        default=[],
+        help="Declare output plot file (do not put into DVC cache).",
```
Comment (efiop) [inline dvc/command/run.py:158]: Just copying help from the options above. Clearly this is not great.

File: dvc/serialize.py [hunk 5]
```diff
@@ -35,10 +36,10 @@ def _get_out(out):
     res = OrderedDict()
     if not out.use_cache:
         res[PARAM_CACHE] = False
-    if out.metric:
-        res[PARAM_METRIC] = True
-    elif out.persist:
```
Comment (efiop) [inline dvc/serialize.py:40]: These are not really mutually exclusive, but our current schema for dvc.yaml doesn't support it. Plus our old dvcfiles were fine with it, so maybe worth just allowing it. Will revisit this line in this PR.

File: tests/func/plots/test_plots.py [hunk 6]
```diff
@@ -27,10 +27,14 @@ def _remove_whitespace(value):
     return value.replace(" ", "").replace("\n", "")
 
 
-def _run_with_metric(tmp_dir, metric_filename, commit=None, tag=None):
-    tmp_dir.dvc.run(metrics_no_cache=[metric_filename], single_stage=True)
```
Comment (efiop) [inline tests/func/plots/test_plots.py:31]: Will need to revisit this helper at some point. It was (and is) a bit hackish.

**Comment by Suor** [comment]
Do we plan to add default metric in a plot there? Current "strategy" of selecting a last key doesn't look well.

**Comment by efiop** [comment]
@Suor Could you elaborate?

**Comment by Suor** [comment]
@efiop plot data file contains several plots, i.e. values of several metrics over something. [Here](https://github.com/iterative/johnnie-metrics) it's `plots.csv` with several columns for example. When `plot diff` shows the plot it selects, which column to use by selecting last column. This is not a good heuristic most time.

I think we should have a way to set default metric to build plot from, probably withing a dvc file.

**Comment by pared** [comment]
@Suor 
We discussed with @efiop that for starters, we should let the user point in stage whether the given metric is plot or not. You are right that the particular parameters should be included, but we should probably do that as next "stage" of this issue.

**Comment by Suor** [comment]
Adding to the above, there is a `-s`, `--select` option to do that on command line. That one accepts either field name or jsonpath.

**Comment by Suor** [comment]
@pared knowing default metric to select from plot data would be very valuable to the corresponding viewer feature. Now there is no interface to change anything, so default matters, it will matter even when some interface will be added. Will also improve cli experience.

**Comment by pared** [comment]
@Suor to not extend this PR scope, maybe we should create p0 for that?

**Comment by Suor** [comment]
@pared I have no preference on where it should go. Let @efiop decide whether he wants to bundle or separate it.

**Comment by Suor** [comment]
Another issue I see is that default metric (`y` axis) is thing across commits, but the `dvc` is stored within commit. So where do we get that? From master, some of the commits involved?

File: dvc/command/plots.py [hunk 7]
```diff
@@ -44,10 +43,9 @@ def run(self):
             else:
                 fields = set(self.args.select.split(","))
         try:
-            plot_string = self.repo.plot(
+            plot_string = self._func(
```
Comment (pared) [inline dvc/command/plots.py:46]: It is no longer a string, right?

File: dvc/output/local.py [hunk 8]
```diff
@@ -71,17 +71,18 @@ def dumpd(self):
         return ret
 
     def verify_metric(self):
-        if not self.metric:
+        if not self.metric or self.plot:
```
Comment (skshetry) [inline dvc/output/local.py:74]: Is this correct? Or, should it be `not self.plot`?

File: dvc/serialize.py [hunk 9]
```diff
@@ -35,10 +36,10 @@ def _get_out(out):
...
+    if out.persist:
         res[PARAM_PERSIST] = True
+    if out.plot and isinstance(out.plot, dict):
+        res.update(out.plot)
```
Comment (skshetry) [inline dvc/serialize.py:42]: Can you please point me where this is being set as dict? I couldn't find it. :confused:

File: dvc/output/local.py (hunk 8 above)
Comment (efiop) [reply dvc/output/local.py:74]: It is correct. Checks for plots don't differ from metric.

File: dvc/serialize.py (hunk 9 above)
Comment (efiop) [reply dvc/serialize.py:42]: @skshetry It is actually not set anywhere programmatically (we might introduce `dvc plots modify` later, but for now it should be set by-hand).

**Comment by efiop** [comment]
@Suor Great questions! For now going only with `templates`, but we'll definitely introduce additional options for it (like labels and stuff) later.

**Comment by efiop** [comment]
Ok, merging for now. Will have a followup PR.

**Pull request merged**